src/riot-rs-threads/src/arch/mod.rs

```diff
@@ -1,17 +1,65 @@
+/// Arch-specific implementations for the scheduler.
+pub trait Arch {
+    const DEFAULT_THREAD_DATA: Self::ThreadData;
+
+    type ThreadData;
+
+    /// Sets up the stack for newly created threads and returns the sp.
+    ///
+    /// After running this, the stack should look as if the thread was
+    /// interrupted by an ISR.
+    ///
+    /// It sets up the stack so when the context is switched to this thread,
+    /// it starts executing `func` with argument `arg`.
+    /// Furthermore, it sets up the link-register with the [`crate::cleanup`] function that
+    /// will be executed after the thread function returned.
+    fn setup_stack(stack: &mut [u8], func: usize, arg: usize) -> usize;
+
+    /// Trigger a context switch.
+    fn schedule();
+
+    /// Setup and initiate the first context switch.
+    fn start_threading(next_sp: usize);
+
+    /// Manually write the pointers to thread data into the registers, as ABI would return the values via stack.
```

Comment:
While the functions above all make sense to have them in the trait, I'm not sure about the ones below.

`return_data_in_regs()` and `wfi()` are *very* architecture specific. (`wfi()` also overloads `cortexm::asm::wfi()`).
I know they're what's needed to keep `sched()` platform independent.
But, it might make more sense to just describe the invariants of the context switching process that's externally visible and depended on, like these two:

```rust
threads.threads[current_pid as usize].sp = old_s;
threads.current_thread = Some(next_pid);
```

... and leave the exact `sched()` machinery to the architecture.
But, let's see how this evolves once we've added RISC-V and Xtensa. :)

Reply:
Concerning `wfi`: won't we always need some way of sleeping if no thread is ready to run?

And concerning `return_data_in_regs()`: I agree that it is platform specific. I wonder if there is a cleaner way to do this if we also decide for RISC-V that we save the registers in the struct. I'll think about it a bit more, also considering what you wrote in https://github.com/future-proof-iot/RIOT-rs/pull/159#discussion_r1516868257. 
Or if it turns out to be too complicated or hacky, we still have the alternative of using the stack, which would then obsolete this method.

Either way, I'd argue that the majority of the logic in `sched` is platform independent and therefore it make sense to have it extract out of the architecture-specific stuff. Even if the essence of it are just the two lines that you wrote above.

Reply:
> Concerning `wfi`: won't we always need some way of sleeping if no thread is ready to run?

Yup. But, this is version force-enabling the interrupts for a bit, within a critical section, relying on nested interrupts to preempt PendSV. Not sure any other platform will do it this way. :)

Reply:
I'm just saying, the logic in `sched()` is both simple and quite specific. It might make sense to just copy & change for the next platform, and that would be OK, at least for a while. If the abstraction gets too messy, it might not be worth it.

Reply:
> I"ll drop the last commit.

For reference: https://github.com/elenaf9/RIOT-rs/commit/ccfacfa4fc028d4a98ecfc992ee913af66008470 was the commit that implemented this changes and that I've now dropped again.